test/BucketAuction.test.ts

```diff
@@ -50,93 +51,444 @@ describe('BucketAuction', function () {
+        expect(await readonlyConn.refundAmount(readonly.address)).to.be.equal(run.refund);
+
+        await readonlyConn.claimTokensAndRefund();
+        const userData = await readonlyConn.getUserData(readonly.address);
+        expect(userData.contribution).to.eq(contribution);
+        expect(userData.tokensClaimed).to.eq(run.numTokens);
+        expect(userData.refundClaimed).to.eq(true);
+
+        balance = (await ownerConn.provider.getBalance(ownerConn.address)).toNumber();
+        expect(balance).to.eq(run.price * run.numTokens);
+      })
+    });
+  })
+
+  describe('User claim and refund', function () {
+    it('user claimTokensAndRefund', async () => {
+      // we then turn on the AuctionActive, so that we can start the bids
+      await ownerConn.setAuctionActive(true);
+
+      // we can make bids
+      await ethers.provider.send('hardhat_setBalance', [
+        readonly.address,
+        ONE_ETH,
+      ]);
+      let balance = (await ownerConn.provider.getBalance(ownerConn.address)).toNumber();
+      expect(balance).to.eq(0);
+      await expect(readonlyConn.bid({value: 100})).to.emit(readonlyConn, 'Bid');
+
+      balance = (await ownerConn.provider.getBalance(ownerConn.address)).toNumber();
+      expect(balance).to.eq(100);
+
+      // and then we prepare to close the auction and settle the price and refund
+      await ownerConn.setAuctionActive(false);
+      await ownerConn.setPrice(80);
+
+      expect(await ownerConn.getClaimable()).to.be.false;
+      await ownerConn.setClaimable(true);
+      expect(await ownerConn.getClaimable()).to.be.true;
+
+      await expect(readonlyConn.claimTokensAndRefund()).to.emit(readonlyConn, 'Transfer');
+      const userData = await readonlyConn.getUserData(readonly.address);
+      expect(userData.contribution).to.eq(100);
+      expect(userData.tokensClaimed).to.eq(1);
+      expect(userData.refundClaimed).to.eq(true);
+      balance = (await ownerConn.provider.getBalance(ownerConn.address)).toNumber();
+      expect(balance).to.eq(80);
+
+      // we cannot claim again after the first successful claim
+      await expect(readonlyConn.claimTokensAndRefund()).to.be.revertedWith('UserAlreadyClaimed');
+
+      // withdraw
+      await expect(ownerConn.withdraw()).to.emit(ownerConn, 'Withdraw');
+      balance = (await ownerConn.provider.getBalance(ownerConn.address)).toNumber();
+      expect(balance).to.eq(0);
+    });
+
+    it('Reverts if price not set', async () => {
+      await ownerConn.setAuctionActive(true);
+
+      await ethers.provider.send('hardhat_setBalance', [
+        readonly.address,
+        ONE_ETH,
+      ]);
+      await expect(readonlyConn.bid({value: 100})).to.emit(readonlyConn, 'Bid');
+
+      await ownerConn.setAuctionActive(false);
+      await ownerConn.setClaimable(true);
+
+      await expect(readonlyConn.claimTokensAndRefund()).to.be.revertedWith('PriceNotSet');
+    });
+
+    it('Reverts if not claimable', async () => {
```

Comment:
can we add one test that

- user A try to claim but user A didn't make any bids before, trying to see what error it should revert

Reply:
Added the case.

There's no error thrown out, we just not issue refund nor send tokens and we marked the refundClaimed = true for that user.